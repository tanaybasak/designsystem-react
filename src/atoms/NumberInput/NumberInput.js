import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';
import FormHelperText from '../FormHelperText';
import Label from '../Label';
const NumberInput = (
    {
        defaultValue,
        id,
        label,
        disabled,
        ...restProps
    }
) => {
    let [value, setValue] = useState(Number(defaultValue) || 0);
    const inputRef = useRef(null);
    const increment = (event) => {
        event.preventDefault();
        if (inputRef.current.stepUp) {
            inputRef.current.stepUp();
        } else {
            stepUp(inputRef.current, inputRef.current.step === '' ? 1 : inputRef.current.step)
        }

        inputRef.current.focus();
        setValue(inputRef.current.value);
        if (restProps.onChange) {
            restProps.onChange(inputRef.current.value);
        }
    }

    const decrement = (event) => {
        event.preventDefault();

        if (inputRef.current.stepDown) {
            inputRef.current.stepDown();
        } else {
            stepDown(inputRef.current, inputRef.current.step === '' ? 1 : inputRef.current.step)
        }
        inputRef.current.focus();
        setValue(inputRef.current.value);
        if (restProps.onChange) {
            restProps.onChange(inputRef.current.value);
        }
    }

    const stepUp = (input, step) => {
        if (input.max !== '' && Number(input.max) < (Number(input.value) + Number(step))) {
            return;
        }
        input.value = Number(input.value) + Number(step);
    }
    const stepDown = (input, step) => {
        if (input.min !== '' && Number(input.min) > (Number(input.value) - Number(step))) {
            return;
        }
        input.value = Number(input.value) - Number(step);
    }

    return (
        <div className={`${prefix}-form-group`} disabled={disabled ? 'disabled' : null} data-invalid={restProps.validationMessage ? true : false}>
            <div className={`${prefix}-form-control ${prefix}-number-input`}>
                <input
                    type="number"
                    className={`${prefix}-form-control`}
                    max={restProps.max ? restProps.max : null}
                    min={restProps.min ? restProps.min : null}
                    step={restProps.step ? restProps.step : null}
                    disabled={disabled ? 'disabled' : null}
                    id={id ? id : null}
                    value={value}
                    onChange={event => {
                        setValue(event.currentTarget.value);
                        if (restProps.onChange) {
                            restProps.onChange(event.currentTarget.value);
                        }
                    }}
                    ref={inputRef}
                />
                <div className={`${prefix}-number-input-control`}>
                    <button className="increment-btn" type="button" tabIndex="-1" onMouseDown={increment} disabled={disabled ? 'disabled' : null}>
                        <svg focusable="false" preserveAspectRatio="xMidYMid meet" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 8 4" aria-hidden="true">
                            <path d="M0 4l4-4 4 4z" />
                        </svg>
                    </button>
                    <button className="decrement-btn" type="button" tabIndex="-1" onMouseDown={decrement} disabled={disabled ? 'disabled' : null}>
                        <svg focusable="false" preserveAspectRatio="xMidYMid meet" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 8 4" aria-hidden="true">
                            <path d="M8 0L4 4 0 0z" />
                        </svg>
                    </button>
                </div>
            </div>

            {label ? <Label htmlFor={id}>{label} </Label> : null}
            {restProps.helperText ? <FormHelperText className="helper-text">{restProps.helperText}</FormHelperText> : null}
            {restProps.validationMessage ? <FormHelperText className="error-msg">{restProps.validationMessage}</FormHelperText> : null}
        </div>
    )
}

NumberInput.propTypes = {
    defaultValue: PropTypes.number,
    id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool
};

NumberInput.defaultProps = {
    defaultValue: 0,
    id: '',
    label: '',
    disabled: false
};

export default NumberInput;
