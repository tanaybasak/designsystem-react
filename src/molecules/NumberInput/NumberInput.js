/* eslint-disable no-empty */
/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';
import FormHelperText from '../../atoms/FormHelperText';
import Label from '../../atoms/Label';
import {
  numberInputMaxValidation,
  numberInputMinValidation,
  numberInputInvalid
} from '../../content';
const NumberInput = ({
  defaultValue,
  id,
  label,
  disabled,
  max,
  min,
  step,
  className,
  helperText,
  required,
  errorMessage,
  ...restProps
}) => {
  let [value, setValue] = useState(Number(defaultValue) || 0);
  let [validationMessage, setValidationMessage] = useState('');
  const inputRef = useRef(null);
  const classnames = `${prefix}-number-input-wrapper ${className.trim()}`;

  useEffect(() => {
    let newValue = Number(value);
    if (required && value === '') {
      setValidationMessage(
        errorMessage && errorMessage.required
          ? errorMessage.required
          : numberInputInvalid
      );
      return;
    } else if (value !== '') {
      if (max != null) {
        if (newValue > max) {
          setValidationMessage(
            errorMessage && errorMessage.max
              ? errorMessage.max
              : `${numberInputMaxValidation} ${max}`
          );
          return;
        }
      }
      if (min != null) {
        if (newValue < min) {
          setValidationMessage(
            errorMessage && errorMessage.min
              ? errorMessage.min
              : `${numberInputMinValidation} ${min}`
          );
          return;
        }
      }
    }
    setValidationMessage('');
  }, [value]);

  const increment = event => {
    event.preventDefault();
    stepUp(
      inputRef.current,
      inputRef.current.step === '' ? 1 : inputRef.current.step
    );
    inputRef.current.focus();
  };

  const decrement = event => {
    event.preventDefault();
    stepDown(
      inputRef.current,
      inputRef.current.step === '' ? 1 : inputRef.current.step
    );
    inputRef.current.focus();
  };

  const stepUp = (input, step) => {
    if (
      input.max !== '' &&
      Number(input.max) < Number(input.value) + Number(step)
    ) {
      return;
    }
    const newValue = Number(input.value) + Number(step) + '';
    setValue(newValue);
    if (restProps.onChange) {
      restProps.onChange(newValue);
    }
  };
  const stepDown = (input, step) => {
    if (
      input.min !== '' &&
      Number(input.min) > Number(input.value) - Number(step)
    ) {
      return;
    }
    const newValue = Number(input.value) - Number(step) + '';
    setValue(newValue);
    if (restProps.onChange) {
      restProps.onChange(newValue);
    }
  };

  // const onNumberInputChange = evt => {
  //   debugger;
  //   const newNumber = evt.target.validity.valid ? evt.target.value : value;
  //   setValue(newNumber);

  //   if (restProps.onChange) {
  //     restProps.onChange(evt.currentTarget.value);
  //   }
  // };

  const onNumberInputKeyDown = evt => {
    const { ctrlKey, metaKey, shiftKey } = evt;
    const key = evt.which || evt.keyCode;
    if (ctrlKey || metaKey || shiftKey) {
    } else if (
      // backspace, delete, arrow keys
      key === 8 ||
      key === 46 ||
      key === 37 ||
      key === 39 ||
      key === 190 ||
      key === 9
    ) {
    } else if (key === 38) {
      increment(evt);
    } else if (key === 40) {
      decrement(evt);
    } else if (key < 48 || key > 57) {
      // non-numeric characters
      evt.preventDefault();
      return;
    }
  };

  return (
    <div className={classnames} disabled={disabled ? 'disabled' : null}>
      {label ? <Label htmlFor={id ? id : null}>{label} </Label> : null}
      {helperText ? (
        <FormHelperText className="helper-text">{helperText}</FormHelperText>
      ) : null}
      <div className={`${prefix}-number-input`}>
        <input
          type="number"
          pattern="^\d*(\.\d+)?$"
          className={`${prefix}-form-control`}
          max={max != null ? max : null}
          min={min != null ? min : null}
          step={step != null ? step : null}
          disabled={disabled ? 'disabled' : null}
          data-invalid={validationMessage !== '' ? true : false}
          id={id ? id : null}
          value={value}
          {...restProps}
          onKeyDown={onNumberInputKeyDown.bind(this)}
          onChange={event => {
            setValue(event.currentTarget.value);
            if (restProps.onChange) {
              restProps.onChange(event.currentTarget.value);
            }
            // console.log('oninput', event);
            // onNumberInputChange(event);
          }}
          ref={inputRef}
        />
        <div className={`${prefix}-number-input-control`}>
          <button
            className="increment-btn"
            type="button"
            tabIndex="-1"
            aria-label="increment button"
            onMouseDown={increment}
            disabled={disabled ? 'disabled' : null}
          >
            <svg
              focusable="false"
              preserveAspectRatio="xMidYMid meet"
              xmlns="https://www.w3.org/2000/svg"
              viewBox="0 0 8 4"
              aria-hidden="true"
            >
              <path d="M0 4l4-4 4 4z" />
            </svg>
          </button>
          <button
            className="decrement-btn"
            type="button"
            tabIndex="-1"
            aria-label="decrement button"
            onMouseDown={decrement}
            disabled={disabled ? 'disabled' : null}
          >
            <svg
              focusable="false"
              preserveAspectRatio="xMidYMid meet"
              xmlns="https://www.w3.org/2000/svg"
              viewBox="0 0 8 4"
              aria-hidden="true"
            >
              <path d="M8 0L4 4 0 0z" />
            </svg>
          </button>
        </div>
      </div>
      {
        <FormHelperText className="error-msg">
          {validationMessage}
        </FormHelperText>
      }
    </div>
  );
};

NumberInput.propTypes = {
  /** Specifies the default value */
  defaultValue: PropTypes.number,
  /** Specifies helper text */
  helperText: PropTypes.string,
  /** Unique Id */
  id: PropTypes.string,
  /** Title for the Number Input */
  label: PropTypes.string,
  /** Disables the  Number Input  */
  disabled: PropTypes.bool,
  /** Specifies the maximum value allowed  */
  max: PropTypes.number,
  /** Specifies the minimum value allowed */
  min: PropTypes.number,
  /** Specifies the number intervals  */
  step: PropTypes.number,
  /** Class/clasess will be applied on the parent div of Number Input  */
  className: PropTypes.string,
  /** Specifies the number is required or not  */
  required: PropTypes.bool,
  /** Used for passing error message  */
  errorMessage: PropTypes.object
};

NumberInput.defaultProps = {
  defaultValue: 0,
  id: '',
  label: null,
  disabled: false,
  max: null,
  min: null,
  step: null,
  helperText: null,
  className: '',
  required: false,
  errorMessage: {
    required: null,
    max: null,
    min: null
  }
};

export default NumberInput;
