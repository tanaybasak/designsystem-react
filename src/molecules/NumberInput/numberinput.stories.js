import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean, number } from '@storybook/addon-knobs';
//@update-path-build-start
import NumberInput from './NumberInput';
//@update-path-build-end
import '../../story.css';
import 'patron-css/dist/patron-style.css';

storiesOf('NumberInput', module).add(
  'basic',
  () => (
    <NumberInput
      defaultValue={10}
      disabled={boolean('Disabled', false)}
      helperText={text('Helper Text', 'Optional Helper text goes here')}
      id="number-input1"
      label={text('Label', 'Number Input')}
      max={number('Max', 100)}
      min={number('Min', 0)}
      onChange={action('Number Change')}
      step={number('Step', 1)}
    />
  ),
  {
    info: {
      text: 'Description About NumberInput Component'
    }
  }
);
