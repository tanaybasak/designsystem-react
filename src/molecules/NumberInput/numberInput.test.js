import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import NumberInput from './NumberInput';

const numberInputComponent = (
    <NumberInput
        defaultValue={10}
        step={1}
        onChange={() => { }}
        id="numberInput2"
        label="Number Input validation"
        helperText="Optional Helper text goes here (max 100 and min 10)"
    />
)

it('Number Input renders correctly', () => {
    const tree = renderer.create(numberInputComponent).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Increment Number', () => {
    const wrapper = mount(numberInputComponent);
    expect(wrapper.find('input').props().value).toEqual(10);
    wrapper.find('.increment-btn').simulate('mouseDown');
    expect(wrapper.find('input').props().value).toEqual(11);
});

it('Decrement Number', () => {
    const wrapper = mount(numberInputComponent);
    expect(wrapper.find('input').props().value).toEqual(10);
    wrapper.find('.decrement-btn').simulate('mouseDown');
    expect(wrapper.find('input').props().value).toEqual(9);
});
