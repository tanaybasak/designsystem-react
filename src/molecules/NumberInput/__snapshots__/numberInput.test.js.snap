// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Input renders correctly 1`] = `
<div
  className="hcl-number-input-wrapper "
  disabled={null}
>
  <label
    className="hcl-label"
    htmlFor="numberInput2"
    onClick={[Function]}
  >
    Number Input validation
     
  </label>
  <div
    className="hcl-helper-text"
    htmlFor=""
    onClick={[Function]}
  >
    Optional Helper text goes here (max 100 and min 10)
  </div>
  <div
    className="hcl-number-input"
  >
    <input
      className="hcl-form-control"
      data-invalid={false}
      disabled={null}
      id="numberInput2"
      max={null}
      min={null}
      onChange={[Function]}
      step={1}
      type="number"
      value={10}
    />
    <div
      className="hcl-number-input-control"
    >
      <button
        aria-label="increment button"
        className="increment-btn"
        disabled={null}
        onMouseDown={[Function]}
        tabIndex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 8 4"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 4l4-4 4 4z"
          />
        </svg>
      </button>
      <button
        aria-label="decrement button"
        className="decrement-btn"
        disabled={null}
        onMouseDown={[Function]}
        tabIndex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 8 4"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0L4 4 0 0z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="hcl-error-msg"
    htmlFor=""
    onClick={[Function]}
  >
    
  </div>
</div>
`;
